<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>Pattern Discovery - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">
<meta name="robots" content="noindex,follow">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-Pattern_Discovery skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">Pattern Discovery</h1>
	<div id="bodyContent">
		<!-- start content -->
<p><br>
</p><p><br>
</p>
<a name="Overview"></a><h2> <span class="mw-headline" id="Overview">Overview</span></h2>
<p>Sequence Pattern Discovery is the process of identifying nucleotide 
or amino acid arrangements, also called motifs, that are enriched in a 
set of sequences. Such motifs may identify regions that have been 
preserved by evolution and which therefore may play a key functional or 
structural role. geWorkbench provides two modes of Sequence Pattern 
Discovery: Regular Discovery and Exhaustive Discovery.
</p><p>Regular Discovery is based on the algorithm SPLASH (Califano, A.,
 2000); it generates a list of all regular expression patterns (motifs) 
that satisfy a user-defined minimum support and a minimum density 
criteria. The former determines the minimum number of times a pattern 
must occur in the sequence set to be reported. This can also be 
expressed as the minimum percent of sequences that must contain the 
pattern. The latter determines how sparse the pattern can be, in other 
words the minimum number of matching characters k (any character except 
for the dot character “.”) over a window of predefined length w.
</p><p>SPLASH-based motif discovery is extremely efficient and can 
process most large protein super-families in a few minutes on a 
conventional workstation. Discovery is uniquely effective in identifying
 sparse patterns using extremely low-density constraints, and the 
results obtained with Discovery can provide the core for a large number 
of more specific local alignments.  
</p><p>Exhaustive Discovery starts from a relatively high minimum 
support (e.g. patterns occurring in 75% of the sequences) and it 
progressively reduces the support, until a statistically significant 
pattern is discovered. Discovered patterns are reported and then masked 
in the sequence set so that they are no longer discovered. Then the 
process continues iteratively until the minimum support reaches a lower 
user-defined limit. Exhaustive Discovery, thus, produces a list of 
non-overlapping motifs in order of support.
</p><p>The Pattern Discovery component is available when a sequence 
(FASTA format) file has been loaded in the Project Folders component, 
and its data node is selected.
</p><p><br>
</p>
<a name="Setting_parameters"></a><h2> <span class="mw-headline" id="Setting_parameters">Setting parameters</span></h2>
<p>A number of parameters can be adjusted by the user to adjust the sensitivity of the search. 
</p>
<a name="Basic_tab"></a><h3> <span class="mw-headline" id="Basic_tab">Basic tab</span></h3>
<p><img alt="PatternDiscovery Params Basic.png" src="patternDiscovery_files/PatternDiscovery_Params_Basic.png" height="318" width="639">
</p><p><br>
Three options for specifying support are available via a pulldown menu:
</p>
<ul><li> <b>Support (Percent of Sequences)</b> - The pattern must appear in at least given percentage of sequences.
</li><li> <b>Support (Number of Sequences)</b> - The pattern must appear in at least the given number of sequences.
</li><li> <b>Support (Number of Occurrences)</b> - The pattern must occur at least the given number of times in the set of sequences (can be more than once per sequence).
</li></ul>
<p><br>
<img alt="PatternDiscovery Params Basic support options.png" src="patternDiscovery_files/PatternDiscovery_Params_Basic_support_options.png" height="216" width="268">
</p><p>Additional options are:
</p>
<ul><li> <b>Minimum Tokens</b> - The minimum number of tokens in a discovered motif.
</li><li> <b>Density Window</b> - A sliding window in which at least the number of tokens set in "Density Window Min. Tokens" must be found.  
</li><li> <b>Density Window Min. Tokens</b> - the minimum number of matching full character tokens (not wildcards) within the "Density Window".
</li></ul>
<a name="Exhaustive_tab"></a><h3> <span class="mw-headline" id="Exhaustive_tab">Exhaustive tab</span></h3>
<p>Parameters specific to the "exhaustive" search algorithm can be set in this tab.
</p><p><img alt="PatternDiscovery Params Exhaustive.png" src="patternDiscovery_files/PatternDiscovery_Params_Exhaustive.png" height="195" width="390">
</p>
<ul><li> <b>Decrement support (%)</b> - sets the size of intervals by 
which support level is decreased in successive searches (default is 5). 
 The decrease is multiplicative, e.g. if one enters 5%, support will be 
reduce to 95% of its previous value at each step.
</li><li> <b>Minimum Support (%)</b> - sets the lower limit on the percentage of sequences that must contain a specific motif (default is 10%).  
</li></ul>
<p><b>Note</b> - Prior to version 2.2.1, if a&nbsp;% sign is included in
 the minimum support text field, the percentage entered will be applied 
to the initial minimum support value.  For example, if the initial 
minimum support is 70%, and "10%" is entered in this field, the final 
stopping value for the calculation will be 7%.
</p>
<ul><li> <b>Minimum Pattern Number</b> - sets a lower limit on the number of motifs in a cluster (point at which support decrease stops).
</li></ul>
<a name="Limits"></a><h3> <span class="mw-headline" id="Limits">Limits</span></h3>
<p><img alt="PatternDiscovery Params Limits.png" src="patternDiscovery_files/PatternDiscovery_Params_Limits.png" height="128" width="343">
</p>
<ul><li> <b>Max. Pattern Number</b> - limits the number of patterns to 
discover.  The actual upper limit to the number of patterns the server 
will return is 99,999.
</li></ul>
<a name="Advanced_tab"></a><h3> <span class="mw-headline" id="Advanced_tab">Advanced tab</span></h3>
<ul><li> <b>Exact Only</b> (default checked)  
<ul><li> When checked, no substitution matrix will be used.  Exact character matches are required.  
</li><li> When checked, the choices for similarity matrix and similarity threshold are disabled.
</li><li> <b>Note</b> - "Exact only" should always be used for DNA as no DNA substitution matrix is provided.
</li><li> When unchecked, the choices for similarity matrix and threshold are enabled.
</li></ul>
</li></ul>
<ul><li> <b>Similarity matrix choice (default BLOSUM50)</b> - Other choices are BLOSUM62 and BLOSUM100.
</li><li> <b>Similarity threshold</b> (default 2) - pairs of amino acids
 with a score higher than the specified threshold in the chosen BLOSUM 
similarity matrix are considered similar.  The input is restricted to 
integers.  (Note that the threshold is however stored as a double).
</li></ul>
<p>The Advanced tab when "Exact Only" is checked:
</p><p><br>
<img alt="PatternDiscovery Params Advanced Exact.png" src="patternDiscovery_files/PatternDiscovery_Params_Advanced_Exact.png" height="196" width="350">
</p><p><br>
The Advanced tab when the "Exact Only" is unchecked:
</p><p><br>
<img alt="PatternDiscovery Params Advanced Matrix.png" src="patternDiscovery_files/PatternDiscovery_Params_Advanced_Matrix.png" height="178" width="322">
</p>
<a name="Run_Pattern_Discovery"></a><h2> <span class="mw-headline" id="Run_Pattern_Discovery">Run Pattern Discovery</span></h2>
<ul><li> Pushing on the button with the curling arrow icon will bring up the session creation box:
</li></ul>
<p><br>
<img alt="PatternDiscovery SessionConnect.png" src="patternDiscovery_files/PatternDiscovery_SessionConnect.png" height="288" width="516">
</p>
<a name="Discovery_Session"></a><h3> <span class="mw-headline" id="Discovery_Session">Discovery Session</span></h3>
<ul><li> <b>Discovery Session Name</b>:  A name is auto-generated for identifying the job on the server, but any name can be entered.
</li><li> <b>Create</b> - Push to start the search.
</li><li> <b>Cancel</b>- Cancel the discovery run.
</li></ul>
<a name="Discovery_Session_Server"></a><h3> <span class="mw-headline" id="Discovery_Session_Server">Discovery Session Server</span></h3>
<ul><li> <b>Server and Port</b>: Columbia supports a Pattern Discovery server at splash.c2b2.columbia.edu, Port 80.
</li><li> <b>Username</b>: Any name can be entered to identify the job.
</li><li> <b>Password</b>: none currently required.
</li></ul>
<p><br>
The progress bar will show the sequence upload and run steps:
</p><p><br>
<img alt="PatternDiscovery ProgressBar.png" src="patternDiscovery_files/PatternDiscovery_ProgressBar.png" height="134" width="326">
</p>
<a name="Viewing_results"></a><h2> <span class="mw-headline" id="Viewing_results">Viewing results</span></h2>
<p>The result of the search can be viewed both in the <b>Pattern Discovery</b> module itself and in sequence viewer modules such as "Sequence" and "Promoter" and "Position Histogram".  In <b>Pattern Discovery</b>
 the results are returned in a table, and the hits for the motif(s) 
selected in the table will be displayed superimposed on the sequences 
displayed e.g. in the Sequence Viewer located above in the Visual area. 
 In this picture, the results were first sorted by the number of tokens.
</p><p>Note that if a substitution matrix was used during discovery, the
 motif may contain a range of possible residues, enclosed by square 
brackets, for a particular position.  This is depicted below under the 
BLOSUM50 result headings.
</p><p>Motifs found are listed in a table with the following columns:
</p>
<ul><li> <b>Hits</b> - The total number of times a motif was found in the sequence dataset, including any multiple hits in individual sequences.  
</li><li> <b>Sequences Hit</b> - The number of sequences which contained at least one occurrence of the discovered motif.
</li><li> <b># of Tokens</b> - the number of full-character tokens in the motif.
</li><li> <b>ZScore</b> - a measure of how often the motif would be 
found in a random set of sequences of the same size and composition as 
the project dataset.
</li><li> <b>Motif</b> - a sequence of tokens, which may be full characterd or wildcard. 
<ul><li> A period (.) represents a wildcard and matches any character.
</li><li> Square brackets are used to indicate multiple possible 
characters at a given position (occurs when a substitution matrix 
(BLOSUM) is used during discovery).
</li></ul>
</li></ul>
<a name="Result_.28Exact_matches.29_sequence_line_view"></a><h3> <span class="mw-headline" id="Result_.28Exact_matches.29_sequence_line_view">Result (Exact matches) sequence line view</span></h3>
<a name="Single_Pattern"></a><h4> <span class="mw-headline" id="Single_Pattern">Single Pattern</span></h4>
<p>Each sequence is shown as a line with length proportional to the sequence length.  All sequences are left-aligned.
</p><p>The position of matches along the sequence are shown using colored boxes (here blue).
</p><p><br>
<img alt="PatternDiscovery Basic histone result exact.png" src="patternDiscovery_files/PatternDiscovery_Basic_histone_result_exact.png" height="606" width="655">
</p><p><br>
If the "All/Matching Pattern" box is checked (red arrow), only sequences that have a match to a selected pattern will be shown.
</p><p><br>
<img alt="PatternDiscovery Basic histone result exact Matching.png" src="patternDiscovery_files/PatternDiscovery_Basic_histone_result_exact_Matching.png" height="233" width="640">
</p>
<a name="Multiple_Patterns"></a><h4> <span class="mw-headline" id="Multiple_Patterns">Multiple Patterns</span></h4>
<p>If more than one pattern is selected, each will be displayed using a 
separate color (note we have scrolled down in the result list, so these 
are not the same hits as above):
</p><p><img alt="PatternDiscovery Basic Histones Result exact 3.png" src="patternDiscovery_files/PatternDiscovery_Basic_Histones_Result_exact_3.png" height="475" width="654">
</p><p><br>
</p>
<a name="Details_as_Hover_Text"></a><h4> <span class="mw-headline" id="Details_as_Hover_Text">Details as Hover Text</span></h4>
<p>If the mouse cursor is placed over a particular pattern match, 
details of any matches at that point are displayed as hover text. 
</p><p>Details include
</p>
<ul><li> current cursor location on sequence
</li><li> pattern(s) matching
</li><li> start and end positions of matching patterns in angle brackets &lt;&gt;.
</li></ul>
<p><img alt="PatternDiscovery Basic Histones Result exact 3 hover.png" src="patternDiscovery_files/PatternDiscovery_Basic_Histones_Result_exact_3_hover.png" height="229" width="680">
</p>
<a name="Result_.28Exact_matches.29_full_sequence_view"></a><h3> <span class="mw-headline" id="Result_.28Exact_matches.29_full_sequence_view">Result (Exact matches) full sequence view</span></h3>
<p>In full sequence view, the actual sequence letters are displayed and matches are again outlined in colored boxes (here blue).  
</p><p><img alt="PatternDiscovery Basic histone result exact seqs.png" src="patternDiscovery_files/PatternDiscovery_Basic_histone_result_exact_seqs.png" height="474" width="657">
</p>
<a name="Result_.28BLOSUM50.29_full_sequence_view"></a><h3> <span class="mw-headline" id="Result_.28BLOSUM50.29_full_sequence_view">Result (BLOSUM50) full sequence view</span></h3>
<p>BLOSUM50 was selected in the Advanced parameters tab and the search 
repeated with the same parameters otherwise.  Note that the 
"All/Matching Pattern" box is checked.  The correspondence between the 
highlighted sequence and the selected regular expression result can be 
seen.
</p><p><img alt="PatternDiscovery Basic histone result blosum50 seqs.png" src="patternDiscovery_files/PatternDiscovery_Basic_histone_result_blosum50_seqs.png" height="748" width="639">
</p>
<a name="Position_Histogram"></a><h3> <span class="mw-headline" id="Position_Histogram">Position Histogram</span></h3>
<p>The Position Histogram displays (binned) support of each motif along the length of the sequences.
</p>
<ul><li> Support - The number of times a motif is found among all the 
sequences at a particular position or within a particular range of 
positions (bin step size) is counted.  The result is displayed as 
fractional support for the motif at that position.
</li></ul>
<ul><li> <b>Step</b> - the sequences are divided into bins of the specified "step" size.  The step size is entered as an integer.
</li><li> <b>Plot Position</b> - push to draw a new histogram.
</li><li> <b>Image snapshot</b> - place a copy of the histogram as an image into the Project Folders component.
</li></ul>
<p><br>
</p><p>Note that this will only be useful to the extent that the sequences are in some way aligned before analysis.
</p><p>The figure below shows three motifs selected.  Each is displayed in a unique color in the position histogram.
</p><p><img alt="PatternDiscovery Basic Histones Result exact 3 histogram.png" src="patternDiscovery_files/PatternDiscovery_Basic_Histones_Result_exact_3_histogram.png" height="523" width="656">
</p>
<a name="Adding_results_to_the_Projects_Folder"></a><h2> <span class="mw-headline" id="Adding_results_to_the_Projects_Folder">Adding results to the Projects Folder</span></h2>
<p>The results of a run of <b>Pattern Discovery</b> are automatically placed in the Project Folder:
</p><p><br>
<img alt="PatternDiscovery Result Node.png" src="patternDiscovery_files/PatternDiscovery_Result_Node.png" height="152" width="272">
</p>
<a name="Options_to_Save_or_Mask_Patterns"></a><h2> <span class="mw-headline" id="Options_to_Save_or_Mask_Patterns">Options to Save or Mask Patterns</span></h2>
<p>Several operations are possible on the returned patterns.  The 
options menu can be seen by right-clicking on a selection of one or more
 returned patterns.
</p><p>1. The patterns can be saved, either with their positions on the original sequences, or just as regular expressions.
</p><p>2. The patterns can be masked out of the query sequence.
</p><p>The options shown in the picture below are:
</p>
<ul><li> <b>Mask Pattern</b> - The selected pattern(s) will be masked out of the sequence for future searches.  They will not be re-discovered.
</li></ul>
<ul><li> <b>Unmask All Patterns</b> - Undo the masking.
</li></ul>
<ul><li> <b>Save Patterns (Regex Only)</b> - This will save the selected pattern(s) in the form of regular expressions, that is, letters and wild-card characters.
</li></ul>
<ul><li> <b>Save Selected Patterns</b> - This will save both the 
selected pattern(s) and their hits to the query sequences.  The 
locations (positions on the query sequences) saved are specific to the 
particular input file used.  The name of this file is saved in the 
pattern file.
</li></ul>
<ul><li> <b>Save All Patterns</b> - This will save both all of the 
patterns and their hits to the query sequences.  The locations positions
 on the query sequences) saved are specific to the particular input file
 used.  The name of this file is saved in the pattern file.
</li></ul>
<p><br>
</p><p><img alt="PatternDiscovery Right click menu.png" src="patternDiscovery_files/PatternDiscovery_Right_click_menu.png" height="171" width="504">
</p>
<a name="Notes_on_display_of_results"></a><h2> <span class="mw-headline" id="Notes_on_display_of_results">Notes on display of results</span></h2>
<ol><li> Pattern Discovery results can only be displayed in the context of the sequences from which they were derived.
</li><li> To reload saved pattern files, see the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Local_Data_Files" title="Local Data Files">| Local Data Files</a> tutorial.
</li><li> There is no provision for filtering out repeats from genomic 
DNA sequence.  Affected sequences should be masked before loading into 
geWorkbench.
</li></ol>
<a name="Example_Pattern_Discovery_runs"></a><h2> <span class="mw-headline" id="Example_Pattern_Discovery_runs">Example Pattern Discovery runs</span></h2>
<a name="Prerequisites"></a><h3> <span class="mw-headline" id="Prerequisites">Prerequisites</span></h3>
<ol><li> Make sure that Pattern Discovery and Position Histogram (if desired) are loaded in the Component Configuration Manager.
</li><li> Load a file containing the sequence or sequences to be analyzed into the Project Folders component.
</li></ol>
<p>For this example, we use a file containing a number of histone sequences, <a href="http://wiki.c2b2.columbia.edu/workbench/images/4/49/H1H5_HistoneDB_NHGRI.fasta" class="internal" title="H1H5 HistoneDB NHGRI.fasta"> H1H5_HistoneDB_NHGRI.fasta</a>.
</p><p>The Pattern Discovery component appears in the Command area of 
geWorkbench (lower right quadrant) when a sequence data node is 
selected.
</p>
<a name="Sequence_Selection"></a><h3> <span class="mw-headline" id="Sequence_Selection">Sequence Selection</span></h3>
<p>By default, Pattern Discovery will be run on all sequences in the 
currently selected sequence data node.  However, subsets of sequences 
can be created and activated in the Markers component.  The Pattern 
Discovery component will respect any activated marker sets, restricting 
discovery to those sequences in activated sets.   If no marker sets are 
activated, then all sequences will be used.
</p>
<a name="Setup_.28Exact_matches.29"></a><h3> <span class="mw-headline" id="Setup_.28Exact_matches.29">Setup (Exact matches)</span></h3>
<p>We will try parameters set to allow longer matches.  No changes from 
default are made in the other parameter tabs.  In particular, this 
search uses exact matching of the sequence letters, without 
substitutions.
</p>
<ul><li> Support Percent: 30
</li><li> Min tokens: 10
</li><li> Density Window: 7
</li><li> Density tokens: 4
</li></ul>
<p>These parameters were empirically chosen.  You can try variations to see how it affects the result.
</p><p><br>
<img alt="PatternDiscovery Params Basic histone setup new.png" src="patternDiscovery_files/PatternDiscovery_Params_Basic_histone_setup_new.png" height="318" width="639">
</p><p><br>
</p><p>The results are pictured above in <b>Viewing Results</b>.
</p>
<a name="Setup_.28BLOSUM50.29"></a><h3> <span class="mw-headline" id="Setup_.28BLOSUM50.29">Setup (BLOSUM50)</span></h3>
<p>If the "Exact only" checkbox in the Advanced Parameters tab is 
unchecked, a selected BLOSUM substitution matrix for protein sequences 
will be used.  
</p>
<ul><li> Similarity Threshold - the threshold is input as an integer 
number.  (However, note that it is passed to the Splash server as a 
double).
</li></ul>
<p><img alt="PatternDiscovery Params Advanced Matrix.png" src="patternDiscovery_files/PatternDiscovery_Params_Advanced_Matrix.png" height="178" width="322">
</p><p>The results of repeating the same run as before but using BLOSUM50 are pictured above in <b>Viewing Results</b>.<br>
</p>
<p>
  <a name="References"></a></p>
<h2> <span class="mw-headline" id="References">References</span></h2>
<p>Califano, A. (2000).  SPLASH: structural pattern localization analysis by sequential histograms.  Bioinformatics, Apr;<b>16</b>(4):341-57 (<a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/16/4/341" class="external text" rel="nofollow">link to paper</a>).
</p>
<!-- 
NewPP limit report
Preprocessor node count: 118/1000000
Post-expand include size: 1907/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key workbench:pcache:idhash:1515-0!1!0!!en!2!printable=1 and timestamp 20120309171257 -->
<div class="printfooter">
Retrieved from "<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Pattern_Discovery">http://wiki.c2b2.columbia.edu/workbench/index.php/Pattern_Discovery</a>"</div>
		<div id="catlinks" class="catlinks catlinks-allhidden"></div>		<!-- end content -->
				<div class="visualClear"></div>
	</div>
</div></div>
<div class="visualClear"></div>
<div id="footer">
	<div id="f-poweredbyico"><img src="patternDiscovery_files/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" height="31" width="88"></div>
	<ul id="f-list">
		<li id="lastmod"> This page was last modified on 9 March 2012, at 12:11.</li>
	</ul>
</div>
</div>

