<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>Master Regulator Analysis - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">
<meta name="robots" content="noindex,follow">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-Master_Regulator_Analysis skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">Master Regulator Analysis</h1>
	<div id="bodyContent">
		<!-- start content -->
<p><br>
</p><p><br>
</p>
<a name="Overview"></a><h1> <span class="mw-headline" id="Overview">Overview</span></h1>
<p>Regulatory activity in the context of specific cellular phenotypes 
can be investigated using interaction networks. These are graphs where 
nodes represent genes and an edge between two nodes A and B means that 
genes <i>A</i> and <i>B</i> are participants in the same regulatory activity. E.g., <i>A</i> can be a transcription factor for <i>B</i>; or, <i>A</i> can be an miRNA that silences <i>B</i>. Analysis of such regulatory networks <a href="#Basso2005"> [Basso et al., 2005]</a>
 has convincingly demonstrated their scale-free nature which is 
dominated by a relatively small number of nodes with a large degree of 
connectivity. The genes corresponding to those nodes are known as 
"master regulators" and collectively orchestrate the regulatory program 
of the underlying cellular phenotype(s). 
</p><p>Master Regulator analysis <a href="#Lefebvre2010"> [Lefebvre et al., 2010]</a>
 is an algorithm used to identify transcription factors whose targets 
(e.g., as represented in an ARACNe-generated interactome) are enriched 
for a particular gene signature.  The enrichment is evaluated using a 
statistical test such as Fisher’s exact test.  The objective is to place
 the signature genes within a regulatory context and identify the master
 regulators responsible for coordinating their activity, thus 
highlighting the regulatory apparatus driving phenotypic 
differentiation. 
</p><p>Specifically, given an interaction network <i>I</i>, a (presumed) master regulator gene <i>A</i>, and a set of signature genes, MRA computes the intersection between two sets of genes:
</p>
<ol><li> The neighbors of <i>A</i> in the interaction network <i>I</i> (this gene set is called the <i><b>regulon</b></i> of <i>A</i>).
</li><li> The set of signature genes.  
</li></ol>
<p>Interaction networks are represented as "adjacency matrices".  An 
adjacency matrix lists the connections that each node takes part in, and
 includes a measure of the strength of that interaction (e.g. the mutual
 information in the case of matrices generated by ARACNe).
</p><p>The MRA component supports two methods to evaluate the overlap of
 the signature and the regulon.  Either method will quantify how likely 
it is to encounter an intersection of (at least) the observed size by 
chance alone. A small p-value is taken to imply that gene <i>A</i> may play a significant role in mediating the regulatory program that leads to the differential phenotypes.
</p>
<ul><li> <b>FET Method</b> - this method use Fisher's Exact Test.  This method is implemented locally in geWorkbench.
</li><li> <b>MARINa Method</b> - this method uses GSEA and differs in 
substantial ways from the FET-based method.  This method is only 
implemented as a grid service and currently has restricted availability 
due to its computational cost.
</li></ul>
<p><br>
For the local, FET method, generating a bar-graph display of the effect 
of a master regulator on its regulon requires performing a t-test of 
differential expression on a dataset representing the phenotypes 
distinguished by the signature, ideally the original dataset from which 
the signature was determined.
</p>
<a name="Setting_up_an_MRA_run"></a><h1> <span class="mw-headline" id="Setting_up_an_MRA_run">Setting up an MRA run</span></h1>
<a name="Prerequisites"></a><h2> <span class="mw-headline" id="Prerequisites">Prerequisites</span></h2>
<ul><li> The Master Regulator Analysis (MRA) component must be loaded in the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Component_Configuration_Manager" title="Component Configuration Manager"> Component Configuration Manager</a>.
</li></ul>
<a name="Gene_Expression_dataset"></a><h3> <span class="mw-headline" id="Gene_Expression_dataset">Gene Expression dataset</span></h3>
<p>A gene expression dataset in which the phenotypic signature was 
identified or can be demonstrated.  A t-test of differential expression 
will be run to generate the graphic "bar code" display of the effect of 
the master regulator on its regulon (FET method) or to generate the 
signature gene list (MARINa method).
</p><p><br>
</p>
<a name="Interaction_Network"></a><h3> <span class="mw-headline" id="Interaction_Network">Interaction Network</span></h3>
<p>An interaction network in the form of an adjacency matrix (See <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/File_Formats" title="File Formats">File Formats</a>.
  Networks can be loaded from a file, or calculated with ARACNe from a 
dataset which includes the particular cellular phenotypes being 
investigated.  If calculating the network with ARACNe, all genes to be 
tested as possible master regulators should be used as hubs.
</p>
<a name="Signature_genes_.28FET_method.29"></a><h3> <span class="mw-headline" id="Signature_genes_.28FET_method.29">Signature genes (FET method)</span></h3>
<p>A list of signature gene markers which distinguish between two 
phenotypes.  This list may come from a t-test, clustering, or some 
combination of methods.  The user must define this set using methods 
relevant to the particular dataset and study goals.
</p>
<a name="Candidate_master_regulator_list_.28FET_method.29"></a><h3> <span class="mw-headline" id="Candidate_master_regulator_list_.28FET_method.29">Candidate master regulator list (FET method)</span></h3>
<p>A set of gene markers that will be tested as candidate master 
regulators.  This set may be comprised of e.g. transcription factor and 
signalling pathway genes.
</p><p><br>
</p>
<a name="Parameters_and_Settings"></a><h2> <span class="mw-headline" id="Parameters_and_Settings">Parameters and Settings</span></h2>
<a name="Main"></a><h3> <span class="mw-headline" id="Main">Main</span></h3>
<p>The settings on this tab apply to both the FET and MARINa methods.
</p><p><br>
<img alt="MRA Parameters Main.png" src="MasterRegulatorAnalysis_files/MRA_Parameters_Main.png" height="293" width="1235">
</p>
<a name="Load_Network"></a><h4> <span class="mw-headline" id="Load_Network">Load Network</span></h4>
<p>There are 2 ways to designate the interaction network, represented by
 an adjacency matrix, that will be used for computing the regulons of 
the candidate master regulator genes:
</p>
<ul><li> <b>From File</b>: by choosing a file that describes a network.  
</li><li> <b>From Project</b>: by selecting an adjacency matrix node from the Project Folders component. 
</li></ul>
<a name="Load_Network_from_File"></a><h5> <span class="mw-headline" id="Load_Network_from_File">Load Network from File</span></h5>
<ul><li> The file loading controls will become active when this option is chosen.
</li><li> Press the "Load" button to bring up the file browser.
</li><li> After selecting a file, a second dialog will ask for details about the format and symbols used.
</li></ul>
<p><img alt="MRA Load Network Dialog.png" src="MasterRegulatorAnalysis_files/MRA_Load_Network_Dialog.png" height="144" width="346">
</p><p>File Format:  
</p>
<ul><li> ADJ
</li><li> SIF
</li><li> MARINa 5-column format (internal use only)
</li></ul>
<p>Nodes Represented by: 
</p>
<ul><li> probeset id
</li><li> gene symbol
</li><li> entrez id
</li><li> other
</li></ul>
<p><br>
After the file has been loaded, its name will be displayed in the adjacent text field.
</p><p><br>
</p>
<a name="Load_Network_from_Project"></a><h5> <span class="mw-headline" id="Load_Network_from_Project">Load Network from Project</span></h5>
<p>Several analytical components in geWorkbench (e.g., <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/ARACNe" title="ARACNe"> ARACNe</a>, <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Cellular_Networks_KnowledgeBase" title="Cellular Networks KnowledgeBase"> CNKB</a>)
 produce adjacency matrix results nodes that can be utilized for this 
purpose.  Networks can also be loaded from a file directly into the 
Project.
</p>
<ul><li> The pulldown menu for choosing an available adjacency matrix 
will become active.  Only adjacency matrices that are children of the 
current microarray dataset will be offered.  
</li></ul>
<p>All edges in the network are assumed to be significant, and any strength value included is not used.
</p>
<a name="Enrichment_Threshold"></a><h4> <span class="mw-headline" id="Enrichment_Threshold">Enrichment Threshold</span></h4>
<p>Enter a p-value for the significance at which to accept the overlap 
of the regulon of a candidate TF and the signature set of genes.
For the FET (local service), this is calculated using the FET.  For the 
MARINa (grid service) method, this is calculated using GSEA.
</p>
<a name="FET_.28local_service.29"></a><h3> <span class="mw-headline" id="FET_.28local_service.29">FET (local service)</span></h3>
<p><img alt="MRA Parameters FET.png" src="MasterRegulatorAnalysis_files/MRA_Parameters_FET.png" height="293" width="1235">
</p>
<a name="Master_Regulators"></a><h4> <span class="mw-headline" id="Master_Regulators">Master Regulators</span></h4>
<p>A set of candidate master regulator markers.  
</p>
<ul><li> This set must be loaded into the Markers component before 
running MRA.  The set can be created directly there, or read in from a 
file.
</li></ul>
<a name="Signature_Markers"></a><h4> <span class="mw-headline" id="Signature_Markers">Signature Markers</span></h4>
<p>A set of markers comprising the signature that distinguishes the chosen phenotype from others.  
</p>
<ul><li> This set must be loaded into the Markers component before 
running MRA.  The signature can be generated directly, e.g. through a 
t-test, or loaded from a file.
</li></ul>
<a name="T-test_for_differential_expression"></a><h4> <span class="mw-headline" id="T-test_for_differential_expression">T-test for differential expression</span></h4>
<p>In the Arrays component, a case and a control group must be defined for running a t-test.
</p><p>A "bar-code" graphic is generated using a t-test of differential 
expression.  However, all t-values are accepted (critical alpha = 1) and
 used to order the bars representing the regulon markers.
</p><p>All  that is required is to define sets of arrays representing 
two phenotypes of interest (and distinguished by the signature). At 
least two sets of arrays must be activated, and at least one marked as 
"case", representing the target phenotype of the gene signature.  
"Control" is the default classification.  See also the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Differential_Expression" title="Differential Expression"> Differential Expression tutorial</a>).
</p><p><br>
<img alt="Array set class assignment MRA.png" src="MasterRegulatorAnalysis_files/Array_set_class_assignment_MRA.png" height="400" width="228">
</p><p><br>
</p>
<a name="MARINA_.28grid_service.29"></a><h3> <span class="mw-headline" id="MARINA_.28grid_service.29">MARINA (grid service)</span></h3>
<a name="Overview_2"></a><h4> <span class="mw-headline" id="Overview_2">Overview</span></h4>
<p>The MARINa algorithm runs on a computational cluster at Columbia and 
access is restricted.  It is run from geWorkbench using a grid service.
</p><p><br>
The MARINa algorithm differs substantially from the "FET" method described above.
</p>
<ol><li> The list of signature markers is determined using a t-test in 
the MARINa algorithm, rather than being loaded separately as for the FET
 method.
</li><li> The list of candidate Master Regulators is taken from the network, rather than using a separate list as in the "FET" method.
</li><li> The MARINa method has its own set of parameters.
</li><li> MARINa will use case and control designations of arrays in the
 Arrays component if provided.  However, unlike the FET method, it will 
also run if only control arrays are designated (which is the default 
setting).  In this case, it will divide the arrays into two groups using
 its own method and run the t-test on this internal division of arrays.
</li></ol>
<a name="Selecting_the_MARINa_grid_service"></a><h4> <span class="mw-headline" id="Selecting_the_MARINa_grid_service">Selecting the MARINa grid service</span></h4>
<p>The figure below shows the Services tab when the MRA grid service has been retrieved and selected.
</p><p><img alt="MRA Services Grid Selected.png" src="MasterRegulatorAnalysis_files/MRA_Services_Grid_Selected.png" height="412" width="1235">
</p><p><br>
</p>
<a name="Setting_the_MARINa_Parameters"></a><h4> <span class="mw-headline" id="Setting_the_MARINa_Parameters">Setting the MARINa Parameters</span></h4>
<p>The MARINa algorithm has its own set of parameters, shown here.
</p><p><img alt="MRA Parameters MARINa.png" src="MasterRegulatorAnalysis_files/MRA_Parameters_MARINa.png" height="293" width="1235">
</p>
<a name="Viewing_MRA_analysis_results_.28FET_Method.29"></a><h1> <span class="mw-headline" id="Viewing_MRA_analysis_results_.28FET_Method.29">Viewing MRA analysis results (FET Method)</span></h1>
<p>Following the successful completion of the MRA FET computation, a 
result node (MRA) appears in the Project Folder area of the geWorkbench 
interface, under the microarray experiment node.  Hovering the cursor 
over the MRA result node will show the number of master regulators 
found.
</p><p><br>
<img alt="Project Folders MRA Tooltip.png" src="MasterRegulatorAnalysis_files/Project_Folders_MRA_Tooltip.png" height="162" width="290">
</p><p>The results of the analysis can be visualized in the MRA Viewer component by selecting the result node.
</p>
<a name="MRA_Results_Viewer"></a><h2> <span class="mw-headline" id="MRA_Results_Viewer">MRA Results Viewer</span></h2>
<p>The MRA viewer is structured in 3 distinct areas.
</p><p>(In the figures below, the data is sorted on the "genes in intersection set column").
</p><p><br>
<img alt="MRA viewer GBM FOSL2.png" src="MasterRegulatorAnalysis_files/MRA_viewer_GBM_FOSL2.png" height="494" width="1023">
</p><p><br>
<b>Note</b> - if no significant MRs are found, an empty result node is 
returned to the Project Folders component.  The MRA viewer will appear 
but be empty.
</p>
<a name="Summary_Listing"></a><h3> <span class="mw-headline" id="Summary_Listing">Summary Listing</span></h3>
<p><img alt="MRA Summary listing.png" src="MasterRegulatorAnalysis_files/MRA_Summary_listing.png" height="229" width="530">
</p><p><br>
At upper left in the MRA viewer.  For each candidate master regulator 
found to have a significant effect using Fisher's Exact test, the 
following four columns are displayed:
</p>
<ul><li> <b>Master Regulator</b> - This is either the master regulator 
gene name or the marker/probeset name identifying the corresponding 
array feature (depending on the selection of the radio buttons “Symbol” 
and “Probe set”).
</li><li> <b>FET p-value</b>  -  the p-value from Fisher’s exact test. 
The test utilizes a 2x2 contingency table where rows classify markers as
 belonging to the signature set or not, while columns indicate if a 
marker belongs to the regulon of the master regulator or not. Counts are
 computed using all markers found in the input experiment data. 
(Fischer's exact test includes p-values for more-extreme tables).  
</li><li> <b>Genes in Regulon</b> - the number of markers (genes) found to be first neighbors of the master regulator in the loaded network - its regulon.
</li><li> <b>Genes in Intersection Set</b> - The number of markers found in the intersection of the signature and the regulon of the candidate MR.
</li></ul>
<p>The contents of the table can be ordered by any column, by clicking 
on the column name.  In this figure, the table is sorted by the number 
of genes in the intersection set.  Sorting by p-value gives a similar 
result.
</p><p>Clicking on the radio button for any of the master regulators 
will display the list of intersection genes in a table to the right 
(Detailed Listing), and will draw the regulon bar graph below.
</p><p><br>
</p>
<a name="Export_all_targets"></a><h4> <span class="mw-headline" id="Export_all_targets">Export all targets</span></h4>
<p>This command writes a file to disk containing each MR in the table, along with each MRs targets and the (value) for each target.
</p><p>The master regulators and their markers in the intersection set 
(intersection of each MRs regulon and the signature genes) are exported,
 along with the T-test value calculated for display of the regulon.  
Each master regulator is listed on a line, followed by its intersection 
set markers with their t-test t values.  Each MR is separated by a blank
 line from the preceeding section.  The order in the file is not changed
 by sorting the results table prior to export.
</p><p><b>Export File format</b>:
</p>
<pre>marker, gene name, t-value
</pre>
<p>Example:
</p>
<pre>220462_at, CSRNP3
200660_at, S100A11, 12.541623
201474_s_at, ITGA3, 7.4126143
202910_s_at, CD97, 10.785
....
</pre>
<pre>202614_at, SLC30A9
160020_at, MMP14, 4.415267
200808_s_at, ZYX, 9.006654
200859_x_at, FLNA, 8.309419
....
</pre>
<p><br>
Exported files automatically receive a ".csv" file name extension.
</p>
<a name="Add_Targets_to_Set"></a><h4> <span class="mw-headline" id="Add_Targets_to_Set">Add Targets to Set</span></h4>
<p>Create a new marker set in the Markers component containing the 
intersection set for the selected master regulator.  The set is named 
after the master regulator.
</p>
<a name="Export_Table"></a><h4> <span class="mw-headline" id="Export_Table">Export Table</span></h4>
<p>This command will export the entire master regulator results table to
 a file.  It exports the same information shown on screen, sorted in the
 same way if the table has been sorted on one of the columns.  The user 
can choose to export the table in CSV (.csv) or tab-delimited text 
format (.txt).
</p><p>The following columns are exported:
</p>
<ul><li> Master Regulator	
</li><li> FET P-Value	
</li><li> Genes in regulon (count)
</li><li> Genes in intersection set (count)
</li></ul>
<p><br>
</p><p><br>
</p>
<a name="Detailed_Listing"></a><h3> <span class="mw-headline" id="Detailed_Listing">Detailed Listing</span></h3>
<p>The detailed list shows the genes/markers contained in the intersection set of the  MR regulon and the signature.
</p><p><br>
<img alt="MRA Detailed listing.png" src="MasterRegulatorAnalysis_files/MRA_Detailed_listing.png" height="236" width="481">
</p><p>The genes are displayed in a table with the following columns:
</p>
<ul><li> <b>Genes in intersection set</b>: the names of the genes in the
 intersection set. Either the gene name or the marker/probe set name is 
used (based on the choice of "Symbol" or "Probe Set" radio buttons).
</li><li> <b>-log10(p-value) * sign (t-value)</b>: A modified test 
statistic combining the -log10(p-value) with the sign of the t-value.  
The sign of the t-value indicates positive or negative differential 
expression.
</li></ul>
<p><br>
</p>
<a name="Export_Table_2"></a><h4> <span class="mw-headline" id="Export_Table_2">Export Table</span></h4>
<p>This command will export to a file on disk the contents of the detailed target results table.
</p><p>The file can be written in either tab-delimited (.txt) or CSV (.csv) format.
</p><p>The columns exported are:
</p>
<ul><li> Genes in intersection set
</li><li> -log10(P-value) * sign of t-value
</li></ul>
<p>Example: 
</p>
<pre>S100A11 12.54
FLNA 8.31
ACTA2 7.91
MYL9 7.04
IFITM2 8.03
ITGA5 8.4
</pre>
<a name="Graph_View"></a><h3> <span class="mw-headline" id="Graph_View">Graph View</span></h3>
<p>For a given master regulator <i>A</i> and the intersection between 
its regulon and the set of differentially expressed genes, the graph 
view helps assess if the intersection genes are preferentially 
over-expressed in the cases versus the controls. The biological 
motivation comes from observing <a href="#Lim2009"> [Lim et al., 2009]</a>
 that regulators with multiple targets tend to affect the expression 
level of (most of) their targets in one particular direction: they 
either promote their expression or inhibit it; but they rarely do both 
equally. 
</p>
<ul><li> The red-blue gradient at the bottom of the graph represents the
 range between the lowest (blue)  and the highest (red) t-test statistic
 recorded among all differentially expressed genes. The white area in 
the middle represents zero.
</li><li> The vertical bars correspond to the genes displayed in the 
table under the “Detailed listing” portion of the interface, i.e., the 
intersection between the differentially expressed genes and the regulon 
of the master regulator <i>A</i> currently selected within the “Summary listing” table. 
</li><li> The relative location of a bar on the gradient represents the t-test statistic recorded for the corresponding gene. 
</li><li> Further, the color of each bar provides information about the 
correlation between the expression levels of the target gene and the 
putative master regulator A (correlations are computed as Spearman’s 
correlation, using data from all microarrays in the experiment): red 
means that the two genes are positively correlated (r &gt; 0) while blue
 means that correlation is negative (r &lt; 0).
</li></ul>
<p><br>
FOSL2:
</p><p><img alt="MRA graph GBM FOSL2.png" src="MasterRegulatorAnalysis_files/MRA_graph_GBM_FOSL2.png" height="74" width="955">
</p><p>The bar graph shown above, for FOSL2, indicates that more of its 
regulon genes have positive differential expression in the mesenchymal 
phenotype, and that expression of FOSL2 is positively correlated with 
the differential expression of its regulon in the mesenchymal phenotype.
 
</p><p><br>
ZNF238:
</p><p><img alt="MRA graph GBM ZNF238.png" src="MasterRegulatorAnalysis_files/MRA_graph_GBM_ZNF238.png" height="76" width="959">
</p><p>The bar graph shown above, for ZNF238, indicates that  expression
 of ZNF238 is negatively correlated with the differential expression of 
its regulon in the mesenchymal phenotype.
</p>
<a name="Dataset_History"></a><h1> <span class="mw-headline" id="Dataset_History">Dataset History</span></h1>
<p>Each results node stores the parameter settings used to setup the 
corresponding MRA run. The specific parameter values can be inspected 
within the Dataset History component, after clicking on the MRA results 
node in the Project Folders pane.
</p>
<a name="Example_of_running_MRA_.28FET_Method.29"></a><h1> <span class="mw-headline" id="Example_of_running_MRA_.28FET_Method.29">Example of running MRA (FET Method)</span></h1>
<p>This example uses a dataset comprised of 176 microarrays described in
 Phillips (2006).  The analysis follows that described in Carro et al. 
(2010) for master regulators of Glioblastoma.
</p><p><br>
</p>
<a name="Loading_and_preparing_the_example_data"></a><h2> <span class="mw-headline" id="Loading_and_preparing_the_example_data">Loading and preparing the example data</span></h2>
<a name="Microarray_dataset"></a><h3> <span class="mw-headline" id="Microarray_dataset">Microarray dataset</span></h3>
<ol><li> Load a microarray dataset.  (See <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Tutorial_-_Local_Data_Files" title="Tutorial - Local Data Files" class="mw-redirect"> Local Data Files</a>).
</li><li> When prompted, load the annotation file.
</li></ol>
<a name="Marker_sets"></a><h3> <span class="mw-headline" id="Marker_sets">Marker sets</span></h3>
<p>Load marker sets for:
</p>
<ol><li> the list of candidate master regulators
</li><li> the signature genes.
</li></ol>
<p><img alt="MRA GBM Marker sets.png" src="MasterRegulatorAnalysis_files/MRA_GBM_Marker_sets.png" height="402" width="246">
</p>
<a name="Array_sets"></a><h3> <span class="mw-headline" id="Array_sets">Array sets</span></h3>
<p>Array sets are shown defined for the three phenotypic classes of 
arrays in the dataset: Mesenchymal (MES), Proneural (PN), and 
Proliferative (Prolif).
</p>
<ul><li> MES and PN are "activated" for use in the t-test by checking the boxes next their names.
</li><li> The MES set is classifed as "Case".  Right click on the thumbtack adjacent to the set name.
</li></ul>
<p><img alt="Array set class assignment MRA.png" src="MasterRegulatorAnalysis_files/Array_set_class_assignment_MRA.png" height="400" width="228">
</p>
<a name="Setting_up_the_parameters_and_starting_MRA"></a><h2> <span class="mw-headline" id="Setting_up_the_parameters_and_starting_MRA">Setting up the parameters and starting MRA</span></h2>
<p>In the Project Folders component, right-click on the expression dataset and select "MRA Analysis".
</p><p>In the "Main" parameters tab, 
</p>
<ul><li> <b>Load Network</b> -  load the network, either directly from a file, or choose a network that has been loaded into the Project.
</li><li> <b>P-value</b> - The p-value for the FET may be set as desired.
</li></ul>
<p>If the network is loaded from a file, you will see the following dialog.
</p><p><img alt="MRA Load Network Dialog.png" src="MasterRegulatorAnalysis_files/MRA_Load_Network_Dialog.png" height="144" width="346">
</p><p>Set the network file format (ADJ or SIF) and type of symbol used 
in the file to represent the gene nodes (e.g. marker id, gene symbol, 
Entrez ID).
</p><p><br>
The figure below shows the Main parameters tab after a network has been loaded from file.
</p><p><br>
<img alt="MRA Parameters Main FET Example.png" src="MasterRegulatorAnalysis_files/MRA_Parameters_Main_FET_Example.png" height="242" width="1212">
</p><p><br>
</p><p>In the "FET" parameters tab, select the signature and master regulator marker sets.
</p><p><br>
<img alt="MRA Parameters FET Example.png" src="MasterRegulatorAnalysis_files/MRA_Parameters_FET_Example.png" height="204" width="1216">
</p><p><br>
</p>
<ul><li> <b>Master regulators</b> - select the desired set from those loaded in the Markers component.
</li><li> <b>Signature markers</b> - select the desired set from those loaded in the Markers component.
</li></ul>
<p><br>
</p>
<ul><li> Click on the <b>Analyze</b> button.
</li></ul>
<a name="Results"></a><h2> <span class="mw-headline" id="Results">Results</span></h2>
<p>Upon completion of the analysis, an MRA results node is placed in the
 Project Folders tree. The analysis results can be browsed using the MRA
 viewer and are as shown above in the MRA Results Viewer section.
</p>
<a name="References"></a><h1> <span class="mw-headline" id="References">References</span></h1>
<p><span id="Basso2005"></span>
</p>
<ul><li> Basso K, Margolin AA, Stolovitzky G, Klein U, Dalla-Favera R, 
Califano A (2005). Reverse engineering of regulatory networks in human B
 cells. Nat Genet 37(4):382-390 (<a href="http://www.nature.com/ng/journal/v37/n4/abs/ng1532.html" class="external text" rel="nofollow">link to paper</a>).
</li></ul>
<p><span id="Lefebvre2010"></span>
</p>
<ul><li> Carro MS, Lim WK, Alvarez MJ, Bollo RJ, Zhao X, Snyder EY, 
Sulman EP, Anne SL, Doetsch F, Colman H, Lasorella A, Aldape K, Califano
 A, Iavarone A  (2010)  The transcriptional network for mesenchymal 
transformation of brain tumors.  Nature 463(7279):318-25.
</li><li> Lefebvre C, Rajbhandari P, Alvarez MJ, Bandaru P, Lim WK, Sato
 M, Wang K, Sumazin P, Kustagi M, Bisikirska BC, Basso K, Beltrao P, 
Krogan N, Gautier J, Dalla-Favera R, Califano A (2010)  A human B-cell 
interactome identifies MYB and FOXM1 as master regulators of 
proliferation in germinal centers.  Mol Syst Biol.  6:377. PMID: 
20531406 (<a href="http://www.ncbi.nlm.nih.gov/pubmed/20531406" class="external text" rel="nofollow">link to paper</a>).
</li></ul>
<p><span id="Lim2009"></span>
</p>
<ul><li> Lim WK, Lyashenko E, Califano A: Master regulators used as 
breast cancer metastasis classifier. Pac Symp Biocomput. 2009:504-15 (<a href="http://psb.stanford.edu/psb-online/proceedings/psb09/lim.pdf" class="external text" rel="nofollow">link to paper</a>).
</li><li> Phillips HS, Kharbanda S, Chen R, Forrest WF, Soriano RH, Wu 
TD, Misra A, Nigro JM, Colman H, Soroceanu L, Williams PM, Modrusan Z, 
Feuerstein BG, Aldape K (2006)  Molecular subclasses of high-grade 
glioma predict prognosis, delineate a pattern of disease progression, 
and resemble stages in neurogenesis.  Cancer Cell 9(3):157-73.
</li></ul>

<!-- 
NewPP limit report
Preprocessor node count: 161/1000000
Post-expand include size: 1907/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key workbench:pcache:idhash:1883-0!1!0!!en!2!printable=1 and timestamp 20120309002758 -->
<div class="printfooter">
Retrieved from "<a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Master_Regulator_Analysis">http://wiki.c2b2.columbia.edu/workbench/index.php/Master_Regulator_Analysis</a>"</div>
		<div id="catlinks" class="catlinks catlinks-allhidden"></div>		<!-- end content -->
				<div class="visualClear"></div>
	</div>
</div></div>
<div class="visualClear"></div>
<div id="footer">
	<div id="f-poweredbyico"><img src="MasterRegulatorAnalysis_files/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" height="31" width="88"></div>
	<ul id="f-list">
		<li id="lastmod"> This page was last modified on 24 February 2012, at 19:01.</li>
	</ul>
</div>
</div>

