<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>CaArray - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">
<meta name="robots" content="noindex,follow">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-CaArray skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">CaArray</h1>
	<div id="bodyContent">
		<!-- start content -->
<p><br>
</p><p><br>
</p>
<a name="The_caArray_Open_File_dialog"></a><h1> <span class="mw-headline" id="The_caArray_Open_File_dialog">The caArray Open File dialog</span></h1>
<p>geWorkbench can retrieve gene expression data from remote instances 
of the NCICB's caArray database.  These may be copies maintained by the 
NCI itself, or copies maintained locally at your or another institution.
  You can maintain settings for any number of different caArray 
installations.
</p>
<a name="Recent_Changes"></a><h2> <span class="mw-headline" id="Recent_Changes">Recent Changes</span></h2>
<ul><li> The caArray 2.5.0 communication client code is not backward 
compatible with that used in caArray 2.4 or earlier versions.  With the 
release of geWorkbench 2.4.0, only communication with caArray 2.5 
installations will be directly supported.
</li></ul>
<ul><li> With geWorkbench 2.4.0, this dialog was renamed from "Remote" to "caArray 2.5" to more accurately reflect its current function.
</li></ul>
<ul><li> As of release 2.3.0, improvements in memory management allow 
large numbers of arrays to be downloaded.  Downloads of more than 500 
arrays at one time have been tested without incident.
</li></ul>
<ul><li> Multiple arrays downloaded at the same time are automatically merged together into a single data set.
</li></ul>
<p><br>
</p>
<a name="Connecting_with_caArray"></a><h2> <span class="mw-headline" id="Connecting_with_caArray">Connecting with caArray</span></h2>
<p>To connect to a remote caArray, right-click on <b>Project</b> in the geWorkbench main window,
</p><p><br>
<img alt="T ProjectFolders OpenFiles.png" src="caArray_files/T_ProjectFolders_OpenFiles.png" height="277" width="393">
</p><p><br>
which will bring up the <b>Open File(s)</b> dialog.   Click the <b>caArray 2.5</b> radio button.   The <b>Open File</b> dialog window will be expanded to include remote caArray sources as shown here.
</p><p><br>
<img alt="Open File caArray.png" src="caArray_files/Open_File_caArray.png" height="375" width="544">
</p><p><br>
</p>
<a name="Connection_and_Query_Controls"></a><h2> <span class="mw-headline" id="Connection_and_Query_Controls">Connection and Query Controls</span></h2>
<p>The buttons at the bottom of the remote file dialog are:
</p>
<ul><li> <b>caArray (Source)</b> menu - Shows a list of caArray 
instances that have been configured.  Entries for the "Production" and 
"Stage" instances of caArray maintained by NCI are preconfigured.
</li><li> <b>Go</b> button - downloads a list of all available experiments from the remote source.
</li><li> <b>Filtering</b> - The list of available experiments can be "filtered" to show only those matching particular criteria. 
</li></ul>
<p><img alt="T OpenFile Remote Filtered human plus.png" src="caArray_files/T_OpenFile_Remote_Filtered_human_plus.png" height="268" width="622">
</p>
<ul><li> The available filtering options are:
<ul><li> <b>Categories</b>: Experiment
</li><li> <b>Field Selection</b>: Array Provider, Organism, Principal Investigator
</li><li> <b>Values</b>:
<ul><li> <b>Array Provider</b>: Affymetrix, Agilent, GenePix, Illumina, ImaGene, Niblegen, ScanArray, UCSF Spot.
</li><li> <b>Organism</b>: many entries, including human, mouse, fly etc...
</li><li> <b>Principal Investigator</b>: PIs of listed public experiments in the particular instance of caArray.
</li></ul>
</li></ul>
</li><li> <b>Add A New Profile</b> button - Opens the Data Source Definition Page used to add a new instance of caArray.
</li><li> <b>Edit Profile</b> button - Edit the currently selected profile.
</li><li> <b>Delete Profile</b> - Remove the currently selected profile.
</li></ul>
<a name="Experiment_Selection_and_Download_Controls"></a><h2> <span class="mw-headline" id="Experiment_Selection_and_Download_Controls">Experiment Selection and Download Controls</span></h2>
<ul><li> <b>caArray Experiments</b> - This pane will contain the list of expriments retrieved from a particular instance of caArray.
</li><li> <b>Experiment Information</b> - clicking on any experiment in the list will show its details here.
</li><li> <b>Number of Assays</b> -  The number of hybridizations for 
which data is available in this experiment, only available after "Show 
Arrays" has been pushed.
</li><li> <b>Show Arrays</b> - Retrieve the full list of available hybridizations for this the selected experiment.
</li><li> <b>Open</b> - Download the hybridizations which have been selected in the list.
</li><li> <b>Cancel</b> - Dismiss the Remote dialog.
</li></ul>
<a name="Loading_data_from_an_instance_of_caArray"></a><h1> <span class="mw-headline" id="Loading_data_from_an_instance_of_caArray">Loading data from an instance of caArray</span></h1>
<a name="Setting_up_the_connection"></a><h2> <span class="mw-headline" id="Setting_up_the_connection">Setting up the connection</span></h2>
<p>On the <b>Open File</b> dialog, you can use
</p>
<ul><li> <b>Add a New Profile</b>  to set up a new connection to an instance of caArray.  Use
</li><li> <b>Edit Profile</b> to change the settings of an existing profile. 
</li></ul>
<p>The settings for the production instance of caArray at the NCI are shown here:
</p><p><img alt="T OpenFile Remote Edit Profile.png" src="caArray_files/T_OpenFile_Remote_Edit_Profile.png" height="396" width="255">
</p>
<ul><li> <b>Profile Name</b> - assigns a name to the profile
</li><li> <b>Protocol</b> - Method by which to communicate with the remote caArray server.
<ul><li> <b>HTTP</b>  - this is the only protocol currently supported.
</li><li> <b>HTTPS</b> - not used.
</li><li> <b>RMI</b> - not used.
</li></ul>
</li><li> <b>Hostname</b> - the URL for the desired caArray server.
</li><li> <b>Port</b> - the port on the remote caArray server at which caArray is available.
</li><li> <b>Username</b> and <b>Password</b> (optional) - caArray 
supports retrieval of data from private experiments.  These fields allow
 the user to provide his or her caArray credentials to gain access to 
any private experiments to which he or she has rights.  No username or 
password is needed for accessing public experiments.
<ul><li> <b>Note</b> -  Once a username and password have been entered 
and submitted to caArray, you cannot go back to using no 
username/password, except by restarting geWorkbench.  However you can 
still put in a different username/password combination.  This is a 
property of the caArray server-side code.
</li></ul>
</li></ul>
<a name="Searching_and_viewing_available_experiments"></a><h2> <span class="mw-headline" id="Searching_and_viewing_available_experiments">Searching and viewing available experiments</span></h2>
<p>If you click on the <b>Go</b> button next to the caArray data source 
at the bottom of the dialog, all available caArray experiments at that 
location will be displayed.  
</p><p>Instead, you can select only particular kinds of experiments by pushing the <b>Filter</b> button.  Here we show experiments of type "Human" being selected.
</p><p><br>
<img alt="OpenFile Remote Filtered human.png" src="caArray_files/OpenFile_Remote_Filtered_human.png" height="270" width="624">
</p><p><br>
The below figure shows the matching entries in the database.  
</p><p><b>Experiment ID</b> - Note that the <b>caArray Experiment ID</b>
 of the selected experiment is shown in the upper right-hand corner of 
the dialog.  This same ID will be seen e.g. if you browse caArray 
through its own web interface.
</p><p><br>
<img alt="Open File caArray Human.png" src="caArray_files/Open_File_caArray_Human.png" height="375" width="544">
</p>
<a name="Viewing_the_list_of_arrays_available_in_a_given_experiment"></a><h2> <span class="mw-headline" id="Viewing_the_list_of_arrays_available_in_a_given_experiment">Viewing the list of arrays available in a given experiment</span></h2>
<p>Select an experiment and push the <b>Show Arrays</b> button to see the individual array datasets available for download for that experiment.
</p><p>You can also right-click directly on an experiment and select <b>Show Arrays</b>.
</p><p>(As of geWorkbench v2.2.2, the list of arrays will be displayed sorted alphabetically.)
</p><p><br>
<img alt="Open File caArray show arrays.png" src="caArray_files/Open_File_caArray_show_arrays.png" height="375" width="541">
</p>
<a name="Downloading_select_array_datasets"></a><h2> <span class="mw-headline" id="Downloading_select_array_datasets">Downloading select array datasets</span></h2>
<p>Now we will select four of the arrays of type HG-U133A and push the <b>Open</b> button to begin the download.  Multiple selections will automatically be merged into a single dataset in geWorkbench.
</p><p><br>
<img alt="Open File caArray download selected.png" src="caArray_files/Open_File_caArray_download_selected.png" height="375" width="541">
</p><p><br>
</p><p>You will be prompted to select the quantitation type from those 
available for the experiment.  Although all available quantitation types
 available for the dataset are listed, usually only the signal channel 
should be used.   Here we select CHPSignal (Affymetrix derived data CHP 
file):
</p><p><img alt="T OpenFile Remote QuantiationSelectionChoices.png" src="caArray_files/T_OpenFile_Remote_QuantiationSelectionChoices.png" height="254" width="278">
</p><p><br>
You will also be prompted to load an annotation file, if desired.  Here we choose the Affymetrix HG-U133A annotation file.
</p><p><br>
<img alt="OpenFile Remote Annotation Glioma.png" src="caArray_files/OpenFile_Remote_Annotation_Glioma.png" height="347" width="570">
</p><p><br>
As of geWorkbench 2.4.0, you will also be prompted to specify the type 
of annotation file type.  The two supported annotation file types are
</p>
<ul><li> Affymetrix 3' Expression
</li><li> Affymetrix Gene/Exon 1.0 ST
</li></ul>
<p>Here we select the 3' Expression format.
</p><p><br>
<img alt="Open File Select Annotation Type.png" src="caArray_files/Open_File_Select_Annotation_Type.png" height="126" width="331">
</p><p><br>
</p><p>A progress bar will track the download process:
</p><p><img alt="OpenFile Remote Loading GliomaExpt.png" src="caArray_files/OpenFile_Remote_Loading_GliomaExpt.png" height="134" width="337">
</p><p><br>
</p><p>The resulting data set will appear in the Project Folders component.
The downloaded dataset is name for the experiment name shown in caArray.
</p><p>If multiple arrays were downloaded, they were merged into a 
single dataset, and the individual arrays are listed in the Arrays 
component:
</p><p><br>
<img alt="Open File caArray merged node.png" src="caArray_files/Open_File_caArray_merged_node.png" height="384" width="231">
</p>
<a name="Cannot_merge_different_array_types"></a><h2> <span class="mw-headline" id="Cannot_merge_different_array_types">Cannot merge different array types</span></h2>
<p>geworkbench does not currently support merging data from arrays of 
different types, e.g. HG-U-133A with HG-U133B.  If by accident such a 
merge is requested, the following error will appear:
</p><p><br>
<img alt="OpenFile Remote Open Merge error.png" src="caArray_files/OpenFile_Remote_Open_Merge_error.png" height="157" width="529">
</p>
<!-- 
NewPP limit report
Preprocessor node count: 46/1000000
Post-expand include size: 1847/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key workbench:pcache:idhash:1810-0!1!0!!en!2!printable=1 and timestamp 20120709185544 -->
<div class="visualClear"></div>
<div id="footer">
	<div id="f-poweredbyico"><img src="caArray_files/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" height="31" width="88"></div>
	<ul id="f-list">
		<li id="lastmod"> This page was last modified on 9 July 2012, at 13:54.</li>
	</ul>
</div>
</div>

