<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html dir="ltr" lang="en"><head>
<title>Gene Ontology Term Analysis - Workbench</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="generator" content="MediaWiki 1.16.0">
<meta name="robots" content="noindex,follow">

</head>
<body class="mediawiki ltr ns-0 ns-subject page-Gene_Ontology_Term_Analysis skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content">
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">Gene Ontology Term Analysis</h1>
	<div id="bodyContent">
		<!-- start content -->
<p><br>
</p>
<a name="Overview"></a><h1> <span class="mw-headline" id="Overview">Overview</span></h1>
<p>The <a href="http://geneontology.org/" class="external text" rel="nofollow">Gene Ontology</a>
 project describes genes (gene products) using terms from three 
structured vocabularies: biological process, cellular component and 
molecular function.  
</p><p>A number of analysis methods in geWorkbench produce a list of 
interesting genes, for example, those differentially expressed (t-test, 
ANOVA), or those which show similarities in expression (Hierarchical 
Clustering, SOM, ARACNe).  The Gene Ontology Enrichment component, also 
referred to as the "GO Terms" component, allows the genes in any such 
"changed-gene" list to be characterized using the Gene Ontology terms 
annotated to them.  It asks, whether for any particular GO term, the 
fraction of genes assigned to it in the "changed-gene" list is higher 
than expected by chance (is over-represented), relative to the fraction 
of genes assigned to that term in the reference set.  In statistical 
terms, the analysis tests the null hypothesis that, for any particular 
ontology term, there is no difference in the proportion of genes 
annotated to it in the reference list and the proportion annotated to it
 in the test list. The reference list is typically comprised of all 
genes on a microarray (after any filtering and removal of redundant 
entries).
</p><p>The Gene Ontology (GO Terms) analysis component in geWorkbench is
 built around the Ontologizer 2.0 software product from Peter Robinson's
 group at the Charite Medical Institute of the Humboldt University in 
Berlin.  It provides several methods for over-representation analysis, 
including Term-for-Term, Parent-Child, and Topology.  More information 
about these methods can be found at the Ontologizer website at <a href="http://compbio.charite.de/index.php/ontologizer2.html" class="external free" rel="nofollow">http://compbio.charite.de/index.php/ontologizer2.html</a>, and in the descriptions and references below.
</p><p>The Gene Ontology is structured as a directed acyclic graph 
(DAG).  This has several consequences.  A term can have more than one 
parent, and hence there can be multiple paths from the root by which a 
term can be reached.  The Ontologizer code uses the "true path" property
 of the Gene Ontology in counting genes assigned to a term, by which a 
gene annotated to any term is considered also annotated to all that 
term’s parent terms.  A term may thus show significant 
over-representation through the cumulative effects of its children 
rather than through genes assigned directly to it.
</p><p><br>
</p>
<ul><li> <b>Important Note on Annotation Files</b>
<ul><li> The Gene Ontology Term Analysis component will automatically make use of <b>Affymetrix 3' Expression</b> microarray annotation data if it was loaded along with the microarray dataset.  
</li><li> Although the Affymetrix Human Gene 1.0 ST and Human Exon 1.0 
ST annotation files can be used by geWorkbench, they cannot currently be
 used by the ontology analysis code. 
</li><li> For any other microarray platform type except the Affymetrix 
3' Expression format, the user must supply an alternate gene ontology 
annotation file directly in the Ontologizer setup.  
</li><li> Annotation files obtained from the GO Consortium (www.geneontology.org) can be loaded. 
</li></ul>
</li></ul>
<ul><li> <b>Note</b> - If a marker has an annotation to a GO Term but 
has no gene symbol, it will not be included in the "Reference Gene" list
 or the "Changed Gene" list.
</li></ul>
<a name="Gene_Ontology_OBO_file_source"></a><h1> <span class="mw-headline" id="Gene_Ontology_OBO_file_source">Gene Ontology OBO file source</span></h1>
<p>By default, each time geWorkbench starts, it downloads the latest 
Gene Ontology OBO file from the geneontology.org website.  However, a 
setting in the geWorkbench <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Menu_Bar" title="Menu Bar">Menu Bar</a> Tools item, <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Menu_Bar#Choose_OBO_Source" title="Menu Bar">"Choose OBO Source"</a>, allows an OBO file to be loaded locally from disk instead.  
</p><p><img alt="Tools OBO choose source.png" src="goTermAnalysis_files/Tools_OBO_choose_source.png" height="111" width="352">
</p><p><br>
The file is chosen using a standard file browser.  After the setting has
 been changed, geWorkbench must be restarted before it will take effect.
</p>
<a name="Analysis_Component_GUI"></a><h1> <span class="mw-headline" id="Analysis_Component_GUI">Analysis Component GUI</span></h1>
<a name="Parameters"></a><h2> <span class="mw-headline" id="Parameters">Parameters</span></h2>
<a name="Selection"></a><h3> <span class="mw-headline" id="Selection">Selection</span></h3>
<p><img alt="GeneOntology Analysis Selection.png" src="goTermAnalysis_files/GeneOntology_Analysis_Selection.png" height="349" width="1082">
</p>
<a name="Reference_Gene_List"></a><h4> <span class="mw-headline" id="Reference_Gene_List">Reference Gene List</span></h4>
<p>The first pulldown allows one to choose from the following sources for the reference gene list:
</p>
<ul><li> <b>All Genes</b> - uses all markers in the current microarray dataset.
</li><li> <b>From Set</b> - if chosen, the second pull-down shows the 
available sets defined in the Markers component.  The markers chosen 
will be converted to gene symbols without duplication.
</li><li> <b>From File</b> - if chosen, the "Load" button becomes active.  
</li></ul>
<p><b>Load</b> - If From File is chosen, the user can load a 
comma-separated list of genes to use as the reference set.  Gene symbols
 must be used, not markers (probeset ids).
</p><p><b>Text field</b> - displays the contents of the currently loaded reference gene list, regardless of source.
</p><p>Note - If a marker has an annotation to a GO Term but has no gene symbol, it will not be included in the "Reference Gene" list.
</p>
<a name="Changed_Gene_List"></a><h4> <span class="mw-headline" id="Changed_Gene_List">Changed Gene List</span></h4>
<p>The first pull-down allows one to choose from the following sources for the changed-gene list: 
</p>
<ul><li> <b>From Set</b> - if chosen, the second pull-down shows the 
available sets defined in the Markers component.  The markers chosen 
will be converted to gene symbols without duplication. 
</li><li> <b>From File</b> - if chosen, the "Load" button becomes active.  
</li><li> <b>From Result Node</b> - if chosen, the second pull-down 
shows a list of available differential expression (t-test, ANOVA) result
 nodes from the Project Folders component.
</li></ul>
<p><b>Load</b> - If From File is chosen, the user can load a 
comma-separated list of genes to use as the changed gene list.  Gene 
symbols must be used, not markers (probeset ids).
</p><p><b>Text field</b> - displays the contents of the currently loaded changed-gene list, except if "From Result Node" is chosen.
</p><p>Note - If a marker has an annotation to a GO Term but has no gene symbol, it will not be included in the "Changed Gene" list.
</p>
<a name="Ontology_Selection"></a><h4> <span class="mw-headline" id="Ontology_Selection">Ontology Selection</span></h4>
<p>Not currently implemented, this is intended to allow the loading of 
alternate ontologies besides the three comprising the Gene Ontology.
</p>
<a name="Ontologizer"></a><h3> <span class="mw-headline" id="Ontologizer">Ontologizer</span></h3>
<p><img alt="GeneOntology Analysis Ontologizer2.png" src="goTermAnalysis_files/GeneOntology_Analysis_Ontologizer2.png" height="287" width="723">
</p>
<a name="Annotations"></a><h4> <span class="mw-headline" id="Annotations">Annotations</span></h4>
<p>The annotation file links individual genes and the GO terms they are 
associated with.  The Ontologizer 2.0 code will automatically make use 
of Affymetrix 3' Expression format annotation files.  For other 
microarray platform types, the user will need to construct and/or 
provide an alternate annotation file.  As of geWorkbench release 2.2.1, 
annotation files obtained from  the GO Consortium (www.geneontology.org)
 can be loaded in addition to Affymetrix annotation files.  However, the
 user must ensure that the proper mapping of marker names is used 
between the microarray data and the annotation file.  Please see <a href="http://compbio.charite.de/contao/index.php/howto.html" class="external free" rel="nofollow">http://compbio.charite.de/contao/index.php/howto.html</a> for further information.
</p>
<ul><li> <b>Use loaded annotations</b> - If an Affymetrix 3' Expression 
format annotation file was read in when the microarray dataset was 
loaded, its name is displayed here.
</li><li> <b>Use alternate annotation file</b> - A new or alternate 
annotation file can be entered into the adjacent text field in by 
selecting this option and choosing "Browse".  
</li><li> <b>Browse</b> - this button brings up a file browser for choosing an annotation file.
</li></ul>
<a name="Enrichment_Method"></a><h4> <span class="mw-headline" id="Enrichment_Method">Enrichment Method</span></h4>
<p><img alt="GeneOntology Analysis EnrichmentMethod.png" src="goTermAnalysis_files/GeneOntology_Analysis_EnrichmentMethod.png" height="150" width="359">
</p>
<ul><li> <b>Parent-Child-Union</b> - see <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=17848398&amp;dopt=Abstract" class="external text" rel="nofollow">Grossman et al. (2007)</a> 
</li><li> <b>Parent-Child-Intersection</b> - see <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=17848398&amp;dopt=Abstract" class="external text" rel="nofollow">Grossman et al. (2007)</a> 
</li><li> <b>Probabilistic</b> - this is an experimental method, not yet published.
</li><li> <b>Term-For-Term</b> (default)- see <a href="http://compbio.charite.de/index.php/ontologizer.over.html" class="external text" rel="nofollow">Ontologizer Overrepresentation Analysis</a>
</li><li> <b>Topology-Elim</b> - see <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=16606683&amp;dopt=Abstract" class="external text" rel="nofollow">Alexa et al. (2006)</a> and <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=17098774&amp;dopt=Abstract" class="external text" rel="nofollow">Falcon et al. (2007)</a>
</li><li> <b>Topology-Weighted</b> - see <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=16606683&amp;dopt=Abstract" class="external text" rel="nofollow">Alexa et al. (2006)</a> and <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=17098774&amp;dopt=Abstract" class="external text" rel="nofollow">Falcon et al. (2007)</a>
</li></ul>
<a name="Multiple_Testing_Correction"></a><h4> <span class="mw-headline" id="Multiple_Testing_Correction">Multiple Testing Correction</span></h4>
<p><img alt="GeneOntology Analysis MultipleTesting.png" src="goTermAnalysis_files/GeneOntology_Analysis_MultipleTesting.png" height="175" width="375">
</p>
<ul><li> <b>Benjamini-Hochberg</b> - 
</li><li> <b>Benjamini-Yekutieli</b> - 
</li><li> <b>Bonferroni</b> - 
</li><li> <b>Bonferroni-Holm</b> - 
</li><li> <b>None</b> (default) - 
</li><li> <b>Westfall-Young-Single-Step</b> - 
</li><li> <b>Westfall-Young-Step-Down</b> -
</li></ul>
<a name="Example"></a><h1> <span class="mw-headline" id="Example">Example</span></h1>
<a name="Setup"></a><h2> <span class="mw-headline" id="Setup">Setup</span></h2>
<p>Running a GO Terms analysis requires a list of genes to analyze (the 
study set).  Here, we will run a simple t-test on two classes of 
cell-lines in the BCell-100.exp dataset.
</p><p><br>
</p>
<ol><li> Load Bcell-100.exp with its annotation file e.g. HG_U95Av2.na32.annot.csv.
</li><li> Threshold normalizer: min threshold 1.0.
</li><li> Log2 Normalize
</li><li> In Arrays, select the "Class" list of array sets and activate GC B-cell and GC-Tumor (Case).
</li><li> Run t-test with alpha threshold = 0.01 and using Bonferroni correction.
</li></ol>
<p><br>
A new Marker set is created called "Significant Genes" with 367 markers.
</p><p><br>
The picture below shows the "Significant Genes" set has been chosen for the Changed Gene list.
</p><p><br>
<img alt="GeneOntology Analysis Setup.png" src="goTermAnalysis_files/GeneOntology_Analysis_Setup.png" height="290" width="1063">
</p><p><br>
In the Ontologizer tab, the Enrichment Method used is term-for-term 
(default) and the Bonferroni multiple testing correction is added.
</p><p><img alt="GeneOntology Analysis Setup Ontologizer.png" src="goTermAnalysis_files/GeneOntology_Analysis_Setup_Ontologizer.png" height="261" width="701">
</p>
<a name="Results"></a><h2> <span class="mw-headline" id="Results">Results</span></h2>
<p>The results of running this analysis are shown on the <a href="http://wiki.c2b2.columbia.edu/workbench/index.php/Gene_Ontology_Viewer#GO_Term_Analysis_Results" title="Gene Ontology Viewer"> Gene Ontology Viewer</a> page.
</p>
<a name="References"></a><h1> <span class="mw-headline" id="References">References</span></h1>
<ul><li> Alexa A, Rahnenführer J, Lengauer T (2006). Improved scoring of
 functional groups from gene expression data by decorrelating GO graph 
structure. Bioinformatics <b>22</b>(13), pps. 2600-1607 (<a href="http://www.ncbi.nlm.nih.gov/pubmed/16606683?dopt=Abstract" class="external text" rel="nofollow">link to paper</a>)
</li></ul>
<ul><li> Bauer S, Grossmann S, Vingron M, Robinson PN (2008). 
Ontologizer 2.0--a multifunctional tool for GO term enrichment analysis 
and data exploration. Bioinformatics <b>24</b>(14), pps. 1650-1. (<a href="http://www.ncbi.nlm.nih.gov/pubmed/18511468" class="external text" rel="nofollow">link to paper</a>)
</li></ul>
<ul><li> Falcon S, Gentleman R. (2007) Using GOstats to test gene lists for GO term association. Bioinformatics  <b>23</b>(2), pps. 257-8. (<a href="http://www.ncbi.nlm.nih.gov/pubmed/17098774?dopt=Abstract" class="external text" rel="nofollow">link to paper</a>)
</li></ul>
<ul><li> Grossmann S, Bauer S, Robinson PN, Vingron M (2007). Improved 
detection of overrepresentation of Gene-Ontology annotations with parent
 child analysis. Bioinformatics <b>23</b>(22), pps. 3024-31. (<a href="http://www.ncbi.nlm.nih.gov/pubmed/17848398?dopt=Abstract" class="external text" rel="nofollow">link to paper</a>)
</li></ul>
<ul><li> Robinson PN, Wollstein A, Böhme U, Beattie B. (2004) 
Ontologizing gene-expression microarray data: characterizing clusters 
with Gene Ontology. Bioinformatics <b>20</b>(6), pps. 979-81. (<a href="http://www.ncbi.nlm.nih.gov/pubmed/14764576?dopt=Abstract" class="external text" rel="nofollow">link to paper</a>)
</li></ul>

<!-- 
NewPP limit report
Preprocessor node count: 66/1000000
Post-expand include size: 1847/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key workbench:pcache:idhash:2025-0!1!0!!en!2!printable=1 and timestamp 20120710195916 -->
<div class="visualClear"></div>
<div id="footer">
	<div id="f-poweredbyico"><img src="goTermAnalysis_files/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" height="31" width="88"></div>
	<ul id="f-list">
		<li id="lastmod"> This page was last modified on 10 July 2012, at 14:58.</li>
	</ul>
</div>
</div>

